Port the workflow to ursa

zhichang-guo opened the pull request:
PR Instructions: 
1. Provide details under all headings below. 
2. Assign Clara and one other person as reviewers. 
3. If the PR is not ready for merging, add the "DRAFT/DO NOT MERGE" label.
4. When a PR is ready to merge, remove the "DRAFT/DO NOT MERGE" and email Clara. 
5. Before requesting that the PR be merged, complete the checklist below. 

Notes on preparing PR, using git can be found in README_git

## Describe your changes  
Summarise all code changes included in PR: 
1. Updated the static files, ERA5 forcing data location and file names, changed the input variable name for precipitation;
2. Fixed some glitches which may lead to a failure of the scripts;
3. Added flexibility for selection of resolutions in the namelists for ufs-land-driver and vector2tile;
4. Coded a shell script "build_landDA.sh", which can build all for the workflow;
5. Added some notes for building the workflow and running the land DA experiments; 

List any associated PRs  in the submodules.
N/A

## Issue ticket number and link
List the git Issue that this PR addresses: 
#45

## Test output 
Is this PR expected to pass the DA_IMS_test (ie., does it change the output)? 

Does it pass the DA_IMS_test? 
Yes, it passes the test
If changes to the test results are expected, what are these changes? Provide a link to the output directory when running the test: 
No exchanges to the test results 
## Checklist before requesting a review
- My branch being merged is up to date with the latest develop. 
- I have performed a self-review of my code by examining the differences that will be merged.
- I have not made any unnecessary code changes / changed any default behavior.
- My code passes the DA_IMS_test.

zhichang-guo committed:
port the workflow to URSA

zhichang-guo committed:
Update build_landDA.sh

zhichang-guo committed:
Update README.md

ClaraDraper-NOAA commented on do_submit_test.sh:
```diff
@@ -4,7 +4,9 @@ source settings_cycle_test
 rm -rf ${OUTDIR}
 
-do_submit_cycle.sh settings_cycle_test
+export CYCLEDIR=$(pwd)
```
Why is this necessary?

zhichang-guo replied:
Sometimes the machine can only resolve the absolute path. It broke the workflow once without the change.

tsga replied:
We are explicitly declaring CycleDIR in setting file

ClaraDraper-NOAA commented on README.md:
```diff
@@ -1,3 +1,10 @@
+Script to run cycling DA using GDASApp in cube sphere space, and offline Noah-MP model in vector space.
```
Can you please merge this into the details below? At the moment, there are two different versions given.

zhichang-guo replied:
Yes, I did it

ClaraDraper-NOAA commented on README.md:
```diff
@@ -1,3 +1,10 @@
+Script to run cycling DA using GDASApp in cube sphere space, and offline Noah-MP model in vector space.
+#############################
+1. Download the shell script build_landDA.sh and use the script to clone and build the workflow;
+2. Edit exp_name, STARTDATE, ENDDATE, BASEDIR, RES, and ORES in settings_cycle_test;
```
Move this to after the instruction for running the test case. Please also change the instructions to: 

"Add your experiment details to settings_cycle_template and settings_cycle_DA_template" 

(the list you have here is incomplete).

ClaraDraper-NOAA commented on build_landDA.sh:
```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
```
Please remove all of the git checkout details from the script. 

If you want to include instructions for cloning the GDASApp, then please add these in the readme. This should be something like: 

" The offline workflow requires GDASApp, and is maintained to use the same GDASApp hash as is in the current global_workflow. You can either link a pre-existing version of GDASApp using DA_update/make_links.sh, or clone and build your own version as follows: 

git clone --recursive https://github.com/NOAA-EMC/gdasapp.git GDASApp currentGDASAppHash" 

+ build instructions

zhichang-guo replied:
Yes, removed.

ClaraDraper-NOAA commented on build_landDA.sh:
```diff
@@ -0,0 +1,77 @@
+  cd $new_dir/DA_update/GDASApp
+  ./build.sh -f -v >& build_log.log&
+fi
+echo '--------------------------------'
```
If you want to add a build_all.sh script, you could keep everything below this line (and please change the script name to build_all.sh)

zhichang-guo replied:
Maybe we should drop this script. I just leave this as my personal use.

ClaraDraper-NOAA replied:
Perhaps that's best.

ClaraDraper-NOAA commented on build_landDA.sh:
```diff
@@ -0,0 +1,77 @@
+echo "load modules"
+if [ $machine = 'ursa' ]; then
+cat > $new_dir/land_mods << EOF1
+module purge
```
If you want to switch to writing out the land_mods, please delete the current land_mods file.

ClaraDraper-NOAA commented on build_landDA.sh:
```diff
@@ -0,0 +1,77 @@
+cat > $new_dir/land_mods << EOF1
+module purge
+module use /contrib/spack-stack/spack-stack-1.9.2/envs/ue-oneapi-2024.2.1/install/modulefiles/Core
+module load stack-oneapi/2024.2.1 stack-intel-oneapi-mpi/2021.13 netcdf-fortran/4.6.1
```
This is also where environmental variables specific to each machine would go.

ClaraDraper-NOAA commented on build_landDA.sh:
```diff
@@ -0,0 +1,77 @@
+  echo "compile ufs-land-driver"
+  cd $new_dir/ufs-land-driver
+  if [ $machine = 'ursa' ]; then
+    git pull origin ursa_configure
```
This switches to a different branch. All machines should be able to compile from the same branch - please remove this.

ClaraDraper-NOAA commented on build_landDA.sh:
```diff
@@ -0,0 +1,77 @@
+  cd $new_dir/ufs-land-driver
+  if [ $machine = 'ursa' ]; then
+    git pull origin ursa_configure
+    ./configure 1 #1=ursa-parallel
```
The arguments here need to be set separately for each machine. 

So if $machine=='ursa' 
...
elseif 
...

ClaraDraper-NOAA replied:
Also, the current version of the ufs_land_driver that I have doesn't have URSA=1. Does the land_driver_repo hash to be updated?

ClaraDraper-NOAA commented on build_landDA.sh:
```diff
@@ -0,0 +1,77 @@
+  echo '--------------------------------'
+  echo "compile vector2tile"
+  cd $new_dir/vector2tile
+  cp ../ufs-land-driver/user_build_config .
```
I'm not sure that these are necessarily always the same?

zhichang-guo replied:
I think it should be always the same.

ClaraDraper-NOAA commented on settings_cycle_test:
```diff
@@ -35,7 +35,7 @@ export atmos_forc=era5
 export NPZ=127   # vertical layers
 export ORES=mx100
-export TPATH=/scratch2/NCEPDEV/land/data/fix/20231027/C${RES}/
+export TPATH=/scratch3/NCEPDEV/global/role.glopara/fix/orog/20231027/C${RES}/
```
This is ursa / hera specific. Can you please create an environmental variable, say: 

GFS_FIX=/scratch3/NCEPDEV/global/role.glopara/fix/orog/20231027/

This would then be set per machine, together with the modules.

ClaraDraper-NOAA replied:
I'm also surprised this is not the latest version, which would be 20240917.

zhichang-guo replied:
Mike and I checked: versions 20240917 and 20231027 should be identical.  The script used for the check:  /scratch3/NCEPDEV/land/Zhichang.Guo/Work/test/compare_orog_files.sh

ClaraDraper-NOAA replied:
In that case, can we update to the latest version? If this directory is used elsewhere for other files, I'd like to make sure that the latest version is used.

zhichang-guo replied:
In GFS, and version is set with environment variable, and let the user make the decision and link the orog files.

ClaraDraper-NOAA commented on template.tile2vector:
```diff
@@ -18,7 +18,7 @@
   
 ! Path for static file  
-  static_filename = "/scratch2/NCEPDEV/land/data/forcing/era5/static/CXXRES/ufs-land_CXXRES_hr3_static_fields.nc"
+  static_filename = "/scratch4/NCEPDEV/land/data/ufs-land-driver/vector_inputs/CXXRES.XXORES/ufs-land_CXXRES.XXORES_hr3_static_fields.nc"
```
Can we set /scratch4/NCEPDEV/land/data/ufs-land-driver/vector_inputs/ to a machine-specific environmental variable too. 

Are we still using the hr3 fields?

tsga replied:
Given both hera and ursa are using same directories, and to quickly get to running the (GEFS) tests, maybe we should for now continue use the above scratch4 directory. And defer setting machine specific links for later (We might not even have some of these files in all machines)

Yes, we are still using hr3 files.

ClaraDraper-NOAA commented on template.ufs-noahMP.namelist.era5:
```diff
@@ -4,9 +4,9 @@
-  static_file      = "/scratch2/NCEPDEV/land/data/forcing/era5/static/C96/ufs-land_C96_hr3_static_fields.nc"
-  init_file        = "/scratch2/NCEPDEV/land/data/forcing/era5/init/C96/ufs-land_C96_hr3_era5_init_2009-12-31_23-00-00.nc"
-  forcing_dir      = "/scratch2/NCEPDEV/land/data/forcing/era5/datm/C96/"
+  static_file      = "/scratch4/NCEPDEV/land/data/ufs-land-driver/vector_inputs/CXXRES.XXORES/ufs-land_CXXRES.XXORES_hr3_static_fields.nc"
```
More machine-specific variables.

ClaraDraper-NOAA commented on template.vector2tile:
```diff
@@ -18,7 +18,7 @@
   
 ! Path for static file  
-  static_filename = "/scratch2/NCEPDEV/land/data/forcing/era5/static/CXXRES/ufs-land_CXXRES_hr3_static_fields.nc"
+  static_filename = "/scratch4/NCEPDEV/land/data/ufs-land-driver/vector_inputs/CXXRES.XXORES/ufs-land_CXXRES.XXORES_hr3_static_fields.nc"
```
and again

ClaraDraper-NOAA reviewed (changes requested):
@zhichang-guo Thanks for doing this, it looks good. Please see comments for requested changes.

Two questions: 
1. Is there an update to the ufs-land-driver hash that should be included in this PR? 
2. It looks like we're still using hr3 static file and forcing, and the 2023 orog files. Is this what is being used for GEFS? For the develop workflow, I'd like to update these both to the latest version (we can then revert for GEFS if need be). @barlage For awareness.

zhichang-guo commented:
> @zhichang-guo Thanks for doing this, it looks good. Please see comments for requested changes.
> 
> Two questions:
> 
> 1. Is there an update to the ufs-land-driver hash that should be included in this PR?
> 2. It looks like we're still using hr3 static file and forcing, and the 2023 orog files. Is this what is being used for GEFS? For the develop workflow, I'd like to update these both to the latest version (we can then revert for GEFS if need be). @barlage For awareness.

1. Yes. To my knowledge Mike's PR https://github.com/ufs-community/ccpp-physics/pull/316 has been approved but not merged and the branch ursa_configure of Mike's ufs-land-driver has not been merged yet.
2. We checked. No update has been done to 20231027 version of orog files (identical to version 20240917). We did not bother doing this update. @ClaraDraper-NOAA @barlage

ClaraDraper-NOAA commented:
> > @zhichang-guo Thanks for doing this, it looks good. Please see comments for requested changes.
> > Two questions:
> > 
> > 1. Is there an update to the ufs-land-driver hash that should be included in this PR?
> > 2. It looks like we're still using hr3 static file and forcing, and the 2023 orog files. Is this what is being used for GEFS? For the develop workflow, I'd like to update these both to the latest version (we can then revert for GEFS if need be). @barlage For awareness.
> 
> 1. Yes. To my knowledge Mike's PR [Glacier modifications in NoahMP to address GFSv17 biases ufs-community/ccpp-physics#316](https://github.com/ufs-community/ccpp-physics/pull/316) has been approved but not merged and the branch ursa_configure of Mike's ufs-land-driver has not been merged yet.
> 2. We checked. No update has been done to 20231027 version of orog files (identical to version 20240917). We did not bother doing this update. @ClaraDraper-NOAA @barlage

Can you please then update the hash for the model submodule to what it should be then? (instead of cloning the desired branch in the build script).

ClaraDraper-NOAA commented:
@zhichang-guo Would you be able to push your updates for this today? I'd like to get it merged as soon as we can.

zhichang-guo commented:
> @zhichang-guo Would you be able to push your updates for this today? I'd like to get it merged as soon as we can.

I removed build_landDA.sh, updated ufs-land-driver to Mike's ursa_configure branch. Mike's ccpp-physics branch has not been merged to ufs-community, so I added his ccpp-physics repo as a romote, and updated his branch glacier_mods. Then, I failed to push my updates (error messages say "failed to push some refs"). I do not know how to fix the problem. Also, if we set the machine-specific orog path, how can we broadcast them to the namelists which are to be used by the fortran programs (sed won't work, I think). So I have difficulties to get it done soon. The directory is located in /scratch3/NCEPDEV/stmp/Zhichang.Guo/land-offline_workflow/ @ClaraDraper-NOAA

tsga committed:
Update do_submit_test.sh

tsga reviewed (commented):
@zhichang-guo @ClaraDraper-NOAA It seems you have agreed to exclude/delete the "build_landDA.sh" file from this PR?
If that is the case, the only comments left are 
- the absolute paths to files that I suggest we keep for now
- Clara's comments about changes in the readme. @zhichang-guo can you address those and mark them resolved?

Then we can merge this and start the tests as you pointed out in the email yesterday.

zhichang-guo committed:
Delete build_landDA.sh

zhichang-guo committed:
Update README.md

zhichang-guo commented:
> @zhichang-guo @ClaraDraper-NOAA It seems you have agreed to exclude/delete the "build_landDA.sh" file from this PR? If that is the case, the only comments left are
> 
> * the absolute paths to files that I suggest we keep for now
> * Clara's comments about changes in the readme. @zhichang-guo can you address those and mark them resolved?
> 
> Then we can merge this and start the tests as you pointed out in the email yesterday.

I removed the script build_landDA.sh and mode corresponding changes to the README. Thanks. @tsga

tsga committed:
Update settings_cycle_test

tsga commented:
thank you @zhichang-guo

zhichang-guo commented:
> thank you @zhichang-guo

Nice work, thanks! @tsga
